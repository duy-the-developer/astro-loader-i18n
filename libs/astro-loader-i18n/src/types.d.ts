export type NormalizeLocales<T extends Locales> = {
  [K in keyof T]: T[K] extends string ? T[K] : T[K] extends {
      codes: Array<string>;
  } ? T[K]['codes'][number] : never;
}[number];
