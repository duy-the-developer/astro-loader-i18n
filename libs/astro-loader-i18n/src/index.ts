export { i18nLoader } from "./loaders/i18n-loader";
export { i18nInfileSchema } from "./schemas/i18n-infile-schema";
export { i18nLoaderSchema } from "./schemas/i18n-loader-schema";
export { localePaths } from "./paths/locale-paths";
export { translationPaths } from "./paths/translation-paths";
export { mergeGetStaticPathsResults } from "./utils/merge";
export { createCollection } from "./collections/create-collection";
