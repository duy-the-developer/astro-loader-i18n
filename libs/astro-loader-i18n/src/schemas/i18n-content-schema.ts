import { z } from "astro/zod";

export const localized = <T extends z.ZodTypeAny, Locales extends readonly string[]>(schema: T, locales: Locales) =>
  z.object(
    locales.reduce(
      (acc, key) => {
        acc[key as Locales[number]] = schema;
        return acc;
      },
      {} as Record<Locales[number], T>
    )
  );
