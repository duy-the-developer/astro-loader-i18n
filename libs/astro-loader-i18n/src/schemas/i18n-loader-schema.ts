import { AnyZodObject, z } from "astro/zod";

export const i18nLoaderSchema = z.object({
  translationId: z.string(),
  locale: z.string(),
  contentPath: z.string(),
});

export const extendI18nLoaderSchema = <Z extends AnyZodObject>(schema: Z) => i18nLoaderSchema.merge(schema);

const i18nLoaderCollectionSchema = z.array(
  z.object({
    data: i18nLoaderSchema,
  })
);

export function checkI18nLoaderCollection(obj: unknown): asserts obj is z.infer<typeof i18nLoaderCollectionSchema> {
  const result = i18nLoaderCollectionSchema.safeParse(obj);

  if (!result.success) {
    throw new Error(
      `Invalid collection entry was provided to astro-i18n-loader. Did you forget to use "extendI18nLoaderSchema" to extend the schema in your "content.config.js" definition? Validation failed with:\n\n${result.error}`
    );
  }
}
