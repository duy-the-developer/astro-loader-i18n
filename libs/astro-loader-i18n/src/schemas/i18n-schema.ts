import { z } from "astro/zod";

export const i18nSchema = z.object({
  commonTranslationId: z.string(),
  locale: z.string(),
});
