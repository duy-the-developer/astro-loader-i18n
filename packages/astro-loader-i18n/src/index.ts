export { i18nFolderLoader } from "./loaders/i18n-folder-loader";
export { i18nSchema } from "./schemas/i18n-schema";
