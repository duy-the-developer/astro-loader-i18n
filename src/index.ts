export { i18nFolderLoader } from "./loaders/i18n-folder-loader";
