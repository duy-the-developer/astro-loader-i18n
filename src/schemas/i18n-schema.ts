import { z } from "astro:content";

export const i18nSchema = z.object({
  locale: z.string(),
});
