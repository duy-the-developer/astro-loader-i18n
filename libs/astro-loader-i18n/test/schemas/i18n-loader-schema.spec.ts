import { expect, describe, it } from "vitest";
import { extendI18nLoaderSchema } from "../../src";
import { z } from "astro/zod";
import { checkI18nLoaderCollection } from "../../src/schemas/i18n-loader-schema";

describe("i18nLoaderSchema", () => {
  it("should extend the schema", () => {
    const schema = extendI18nLoaderSchema(z.object({ title: z.string() }));
    expect(JSON.stringify(schema.shape)).toMatchSnapshot();
  });
  it("should throw an error when checkI18nLoaderCollection fails", () => {
    const invalidData = [
      { data: { translationId: "1", locale: "en" } },
      { data: { translationId: "2", locale: "fr" } },
      { data: { translationId: "3" } },
    ];

    expect(() => checkI18nLoaderCollection(invalidData)).toThrowErrorMatchingSnapshot();
  });
});
