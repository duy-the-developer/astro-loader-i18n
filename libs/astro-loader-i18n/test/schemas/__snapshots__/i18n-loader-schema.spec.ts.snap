// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`i18nLoaderSchema > should extend the schema 1`] = `"{"translationId":{"_def":{"checks":[],"typeName":"ZodString","coerce":false},"~standard":{"version":1,"vendor":"zod"}},"locale":{"_def":{"checks":[],"typeName":"ZodString","coerce":false},"~standard":{"version":1,"vendor":"zod"}},"title":{"_def":{"checks":[],"typeName":"ZodString","coerce":false},"~standard":{"version":1,"vendor":"zod"}}}"`;

exports[`i18nLoaderSchema > should throw an error when checkI18nLoaderCollection fails 1`] = `
[Error: Invalid collection entry was provided: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      2,
      "data",
      "locale"
    ],
    "message": "Required"
  }
] 
 Did you forget to use "extendI18nLoaderSchema" to extend the schema?]
`;
